librz/il/rzil_vm.c

```diff
@@ -0,0 +1,724 @@
+ * \param vm, pointer to VM
+ * \param temp, pointer to temporary value
+ */
+RZ_API void rz_il_clean_temp(RzILVM vm, RzILTemp temp) {
+	RZIL_TEMP_TYPE type = temp->type;
+	switch (type) {
+	case RZIL_TEMP_VAL:
+		rz_il_free_value(temp->data);
+		break;
+	case RZIL_TEMP_BOOL:
+		rz_il_free_bool(temp->data);
+		break;
+	case RZIL_TEMP_BV:
+		rz_il_bv_free(temp->data);
+		break;
+	case RZIL_TEMP_EFF:
+		effect_free(temp->data);
+		break;
+	default:
+		break;
+	}
+
+	temp->data = NULL;
+	temp->type = RZIL_TEMP_EMPTY;
+}
+
+/**
+ * Free and clean the temporay values in VM
+ * \param vm RzILVM, pointer to VM
+ */
+RZ_API void rz_il_clean_temps(RzILVM vm) {
+	RzILTemp *temps = vm->temp_value_list;
+	for (int i = 0; i < RZ_IL_VM_MAX_TEMP; ++i) {
+		if (temps[i]) {
+			rz_il_clean_temp(vm, temps[i]);
+		}
+	}
+}
+
+/**
+ * Make a core theory opcode vector
+ * \param num int, number of total opcodes
+ * \param ... op RzILOp, op will be pushed to vector one by one
+ * \return oplist RzPVector*, pointer to the opcode
+ */
+RZ_API RzPVector *rz_il_make_oplist(int num, ...) {
+	va_list args;
+	RzILOp cur_op;
+	RzPVector *oplist = rz_pvector_new((RzPVectorFree)rz_il_free_op);
+
+	va_start(args, num);
+	for (int i = 0; i < num; ++i) {
+		cur_op = va_arg(args, RzILOp);
+		rz_pvector_push(oplist, cur_op);
+	}
+	va_end(args);
+
+	return oplist;
+}
+
+/**
+ * Make a core theory opcode vector, and set id for every opcode
+ * \param id ut64, set id for every opcode in this list
+ * \param num int, number of total opcodes
+ * \param ... op RzILOp, op will be pushed to vector one by one
+ * \return oplist RzPVector*, pointer to the opcode
+ */
+RZ_API RzPVector *rz_il_make_oplist_with_id(ut64 id, int num, ...) {
+	va_list args;
+	RzILOp cur_op;
+	RzPVector *oplist = rz_pvector_new((RzPVectorFree)rz_il_free_op);
+
+	va_start(args, num);
+	for (int i = 0; i < num; ++i) {
+		cur_op = va_arg(args, RzILOp);
+		cur_op->id = id;
+		rz_pvector_push(oplist, cur_op);
+	}
+	va_end(args);
+
+	return oplist;
+}
+
+/**
+ * Store an opcode list to address
+ * \param vm RzILVM, pointer to VM
+ * \param addr RzILBitVector, address of this opcode list
+ * \param oplist RzPVector of RzILOp, core theory opcodes
+ */
+RZ_API void rz_il_vm_store_opcodes_to_addr(RzILVM vm, RzILBitVector addr, RzPVector *oplist) {
+	ht_pp_insert(vm->ct_opcodes, addr, oplist);
+}
+
+/**
+ * Load an opcode list at current pc
+ * \param vm RzILVM, pointer to VM
+ * \return oplist RzPvector of RzILOp, core theory opcodes
+ */
+RZ_API RzPVector *rz_il_vm_load_opcodes_at_pc(RzILVM vm) {
+	return ht_pp_find(vm->ct_opcodes, vm->pc, NULL);
+}
+
+/**
+ * Load an opcode list at address
+ * \param vm RzILVM, pointer to VM
+ * \param addr RzILBitVector, address to load ops
+ * \return oplist RzPvector of RzILOp, core theory opcodes
+ */
+RZ_API RzPVector *rz_il_vm_load_opcodes(RzILVM vm, RzILBitVector addr) {
+	return ht_pp_find(vm->ct_opcodes, addr, NULL);
+}
+
+static void print_val(RzILVal val) {
```

Comment:
Move `print` functions to `librz/core/cil.c`. Put here `RZ_API rz_il_val_as_string()`